### PR TITLE
Alarm input queue

### PR DESCRIPTION
**Pull request by pyhys**
Based on top of #383

### CODE REVIEW

File: src/common/pf_alarm.c [hunk 1]
```diff
@@ -978,12 +978,12 @@ static int pf_alarm_apmx_activate (pnet_t * net, pf_ar_t * p_ar)
             p_ar->apmx[ix].frame_id = PF_FRAME_ID_ALARM_HIGH;
          }
 
-         if (p_ar->apmx[ix].p_alarm_q == NULL)
+         /* Input alarm frame queue */
+         if (p_ar->apmx[ix].alarm_receive_q.p_receive_mutex == NULL)
          {
-            p_ar->apmx[ix].p_alarm_q = os_mbox_create (PNET_MAX_ALARMS);
+            p_ar->apmx[ix].alarm_receive_q.p_receive_mutex = os_mutex_create();
```
Comment (olbjo) [inline src/common/pf_alarm.c:984]: Maybe "mutex" works as name? p_ should be avoid according to coding standard. That it is for the receive queue is given by the containing struct/variable name

**Review by olbjo — COMMENTED** [review]
Great!
I focused on the new alarm part but also randomly looked at other parts. Did not review tests

**Review by olbjo — APPROVED** [review]
👍

File: src/common/pf_alarm.c (hunk 1 above)
Comment (pyhys) [reply src/common/pf_alarm.c]: Done

File: test/test_alarm.cpp [hunk 2]
```diff
@@ -294,60 +294,149 @@ TEST_F (AlarmUnitTest, AlarmCheckAddDiagItemToSummary)
...
+
+TEST_F (AlarmUnitTest, AlarmCheckReceiveQueueHandling)
+{
+   pf_alarm_receive_queue_t queue;
+   pf_apmr_msg_t input_frame;
+   pf_apmr_msg_t output_frame;
+   int ix = 0;
+   int err = 0;
+   const uint16_t POSITION_START_NUMBER = 100;
+   const uint64_t POINTER_START_NUMBER = 1000; // TODO uint64_t not portable
```
Comment (hefloryd) [inline test/test_alarm.cpp:375]: Not sure what this is referring to but note that there is a uintptr_t type capable of holding a pointer on the target platform.

File: src/common/pf_alarm.c [hunk 3]
```diff
@@ -38,8 +38,18 @@
  * for high prio.
  *
  * A frame handler is registered for each of the high prio and low prio
- * alarm frame IDs. The incoming frames are put into high prio and low prio
- * queues (mboxes). The periodic task handles frames in the queues.
+ * alarm frame IDs.
+ *
+ * Each AR has two alarm receive queues, one for low prio and one for high
+ * prio alarm frames. The thread responsible for processing incoming frames
+ * is putting frames into the queues, and another thread (driven by the
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/common/pf_alarm.c:45]: Done

File: src/common/pf_alarm.c [hunk 4]
```diff
@@ -38,8 +38,18 @@
...
+ * periodic task) is retrieving messages from the receive queues.
+ * A message in an input queue contains a pointer to an incoming frame buffer.
+ *
+ * Each AR has its own alarm send queue. As alarms only can be processed
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/common/pf_alarm.c:49]: Done

File: test/test_alarm.cpp (hunk 2 above)
Comment (pyhys) [reply test/test_alarm.cpp:375]: Great, thanks!
Done.

File: src/common/pf_alarm.c [hunk 5]
```diff
@@ -2267,15 +2240,15 @@ static int pf_alarm_send_internal (
    return ret;
 }
 
+/************************ Queue handling ************************************/
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/common/pf_alarm.c:2239]: Now using same logic for both queues

File: src/common/pf_alarm.c [hunk 6]
```diff
@@ -998,12 +981,10 @@ static int pf_alarm_apmx_activate (pnet_t * net, pf_ar_t * p_ar)
...
          }
 
-         if (p_ar->apmx[ix].p_alarm_q == NULL)
-         {
-            p_ar->apmx[ix].p_alarm_q = os_mbox_create (PNET_MAX_ALARMS);
-         }
-         memset (p_ar->apmx[ix].apmr_msg, 0, sizeof (p_ar->apmx[ix].apmr_msg));
-         p_ar->apmx[ix].apmr_msg_nbr = 0;
+         /* Input alarm frame queue */
+         pf_alarm_queue_mutex_create (
```
Comment (hefloryd) [inline src/common/pf_alarm.c:985]: Does the queue need to create/destroy the mutex at runtime? This likely involves malloc/free which is best avoided except for initialization code.

File: src/common/pf_alarm.c [hunk 7]
```diff
@@ -2267,74 +2236,296 @@ static int pf_alarm_send_internal (
...
+
+/**
+ * @internal
+ * Unlock a queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+static void pf_alarm_queue_unlock (pf_queue_accountant_t * p_accountant)
+{
+   assert (p_accountant->mutex != NULL);
```
Comment (hefloryd) [inline src/common/pf_alarm.c:2297]: CC_ASSERT

File: src/common/pf_alarm.c [hunk 8]
```diff
@@ -2267,74 +2236,296 @@ static int pf_alarm_send_internal (
...
+ * @param p_accountant     InOut: Queue accountant
+ * @param p_read_index     Out:   Read index to use
+ * @return 0 if the queue has an item to read (Queue will be locked)
+ *         -1 if the queue is empty (Will not be locked)
+ */
+static int pf_alarm_queue_lock_and_get_readindex (
+   pf_queue_accountant_t * p_accountant,
+   uint16_t * p_read_index)
+{
+   // TODO Remove, find out why it segfaults sometimes.
```
Comment (hefloryd) [inline src/common/pf_alarm.c:2351]: ?

File: test/test_alarm.cpp [hunk 9]
```diff
@@ -294,60 +294,153 @@ TEST_F (AlarmUnitTest, AlarmCheckAddDiagItemToSummary)
...
    int ix = 0;
    int err = 0;
    const uint16_t SEQUENCE_START_NUMBER = 100;
 
    /* Prepare default data */
-   memset (&queue, 0, sizeof (queue));
    memset (&alarm_input, 0, sizeof (alarm_input));
    memset (&alarm_output, 0, sizeof (alarm_output));
-   EXPECT_EQ (queue.count, 0);
+
```
Comment (hefloryd) [inline test/test_alarm.cpp:309]: Check if there is boilerplate setup code here that could be moved to the test fixture.

File: src/common/pf_alarm.c [hunk 10]
```diff
@@ -2267,74 +2236,296 @@ static int pf_alarm_send_internal (
...
+ *         -1 if the queue is full (Will not be locked)
+ */
+static int pf_alarm_queue_lock_and_get_writeindex (
+   pf_queue_accountant_t * p_accountant,
+   uint16_t * p_write_index)
+{
+   os_mutex_lock (p_accountant->mutex);
+   if (p_accountant->count >= PNET_MAX_ALARMS)
+   {
+      os_mutex_unlock (p_accountant->mutex);
```
Comment (hefloryd) [inline src/common/pf_alarm.c:2319]: This function does not always exit with the mutex locked. That leads to confusing code later on, where it looks like calling functions forgot to unlock the mutex. Could the code be organised in a way that avoids the "lock_and"? Once you have allocated the write-index, do you need to keep the lock?

File: src/common/pf_alarm.c (hunk 6 above)
Comment (pyhys) [reply src/common/pf_alarm.c:985]: The queues are part of the big `pf_ar_t` struct, which is `memset()` on connect request from the PLC.
I think changing this should be a separate PR.

File: src/common/pf_alarm.c (hunk 7 above)
Comment (pyhys) [reply src/common/pf_alarm.c]: Done

File: src/common/pf_alarm.c (hunk 8 above)
Comment (pyhys) [reply src/common/pf_alarm.c]: Implemented logic to handle incoming frames before or after the queue is in its initialized state.
Done

File: test/test_alarm.cpp (hunk 9 above)
Comment (pyhys) [reply test/test_alarm.cpp]: There are different types for the arguments in the different test cases, so there is not much gain to move to the fixture.
Clarified this by giving better names.
Done.

File: src/common/pf_alarm.c (hunk 10 above)
Comment (pyhys) [reply src/common/pf_alarm.c]: Restructured the logic.
Done

**Comment by pyhys** [comment]
Renamed queue handling functions to `_post()` and `_fetch()`.

**Pull request merged**